src/ensembl/src/shared/utils/lruCache.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * This is a cache implementing the Least Recently Used algorithm
+ * for preventing it from growing indefinitely. It also can invalidate
+ * individual items based on the max time during which an item is allowed
+ * to be retrieved from the cache.
+ *
+ * The Least Recently Used algorithm requires coordinated use of a doubly linked list
+ * for storing cached values, and a map for quick access to the nodes of the list
+ */
+
+interface LRUCacheInterface {
+  get(key: string): any;
+  set(key: string, value: any): void;
+  reset(): void;
+}
+
+interface LinkedListNodeInterface {
+  value: any;
+  next: LinkedListNode | null;
+  previous: LinkedListNode | null;
+  timestamp: number;
+}
+
+type LRUCacheOptions = {
+  size?: number; // max number of items that can be stored in cache; defaults to 1000
+  maxAge?: number; // time period, in milliseconds, during which a cache item is retrievable
+};
+
+type LinkedListNodeOptions = {
+  key: string;
+  value: any;
+  timestamp: number;
+};
+
+const DEFAULT_SIZE = 1000;
+
+export default class LRUCache implements LRUCacheInterface {
+  private size: number;
+  private maxAge?: number;
+  private currentSize = 0;
+  private head: LinkedListNode | null;
+  private tail: LinkedListNode | null;
+  private listNodesMap: Map<string, LinkedListNode>;
+
+  constructor(options: LRUCacheOptions = {}) {
+    if (options.size && options.size <= 0) {
+      throw new Error('Cache size must be a positive integer');
+    }
+    this.size = options.size || DEFAULT_SIZE;
+    this.maxAge = options.maxAge;
+
+    this.head = null;
+    this.tail = null;
+    this.listNodesMap = new Map();
+  }
+
+  public get(key: string) {
+    const cachedNode = this.listNodesMap.get(key);
+    if (cachedNode) {
+      if (this.isExpiredNode(cachedNode)) {
+        this.removeNode(cachedNode);
+        return;
+      }
+      this.moveNodeToHead(cachedNode);
+      cachedNode.timestamp = Date.now();
+      return cachedNode.value;
+    }
+  }
+
+  public set(key: string, value: any) {
+    const savedNode = this.listNodesMap.get(key);
+    if (savedNode) {
+      this.moveNodeToHead(savedNode);
+      savedNode.value = value;
+    } else {
+      this.addNode(key, value);
+    }
+  }
+
+  public reset() {
+    this.head = null;
+    this.tail = null;
+    this.currentSize = 0;
+    this.listNodesMap.clear();
+  }
+
+  private isExpiredNode(node: LinkedListNode) {
+    if (!this.maxAge) {
+      return false;
+    }
+    return Date.now() - node.timestamp > this.maxAge;
+  }
+
+  private moveNodeToHead(node: LinkedListNode) {
+    const previousNode = node.previous;
+    const nextNode = node.next;
+    if (previousNode) {
+      previousNode.next = nextNode;
+    }
+    if (nextNode) {
+      nextNode.previous = previousNode;
+    }
+    const formerHeadNode = this.head;
+    this.head = node;
+    node.next = formerHeadNode;
+    node.previous = null;
```

Comment:
Shouldn't it be the other way?
```javascript
node.previous = formerHeadNode
node.next = null
```

Reply:
The list is growing towards its tail. The most recently accessed node is the head; the least recently accessed node is the tail.

If you access a node in the middle of the list, you remove it from its place and put it in the head. The head does not have previous nodes, so node.previous is null. The one that used to be head before that becomes less recently accessed, and thus it gets moved towards the tail, which means that the new head node will reference it via node.next.

Reply:
> The list is growing towards its tail.

In the following example, is the node `[null,1,2]` considered to be the head with previous as `null` and next as `2`?
[null,1,2] <-> [1,2,3] <-> [2,3,null]

Reply:
Yes, if I understand your notation correctly.

So when we move node 2 to head, it should be [null,2,1] <-> [2,1,3] <-> [1,3,null].

Uh-oh, did I forget to link node 3 with node 1? I did, didn't I.

Reply:
> Uh-oh, did I forget to link node 3 with node 1? I did, didn't I.

Why will that be required?

Reply:
Well, my list is broken without it:

Given:
[null,1,2] <-> [1,2,3] <-> [2,3,null]

Expected after node switching:
[null,2,1] <-> [2,1,3] <-> [1,3,null]

But am likely getting:
```
[null,2,1] <-> [2,1,3] -> [2,3,null]
           <-----------
```

Reply:
Sorry, now that I am reading through the code of the method again, the only thing that I missed is assigning previous to formerHeadNode.

Starting with:
[null,1,2] <-> [1,2,3] <-> [2,3,null]

If we want to move [1,2,3] to head:

```
    const previousNode = node.previous;
    const nextNode = node.next;
    if (previousNode) {
      previousNode.next = nextNode;
    }
    if (nextNode) {
      nextNode.previous = previousNode;
    }
```
this should establish the mutual links between node 1 and node 3:

```[null,1,3] <-> [1,3,null]```

And the next set of instructions:

```
    const formerHeadNode = this.head;
    this.head = node;
    node.next = formerHeadNode;
    node.previous = null;
```

will move node 2 to head:

```[null,2,1] <-> [null,1,3] <-> [1,3,null]```

So what remains to be done is to set formerHeadNode.previous to node (i.e. to 2):

```[null,2,1] <-> [2,1,3] <-> [1,3,null]```

Reply:
Cool, that sounds right 👍

Reply:
I updated the tests, and found and fixed another bug in the process 😅